### ISSUE TITLE
Exception when starting a stream when no data yet exists (ie. empty Kafka queue)

### ISSUE THREAD

**Issue by cfregly** [brkyvz/streaming-matrix-factorization#1]
@brkyvz 

Hey burak!

I'm having an issue where, upon starting my stream when there is no data yet (ie empty Kafka queue), i'm seeing the following error:

```
java.lang.UnsupportedOperationException: empty collection
    at org.apache.spark.rdd.RDD$$anonfun$reduce$1$$anonfun$apply$36.apply(RDD.scala:1005)
    at org.apache.spark.rdd.RDD$$anonfun$reduce$1$$anonfun$apply$36.apply(RDD.scala:1005)
    at scala.Option.getOrElse(Option.scala:120)
    at org.apache.spark.rdd.RDD$$anonfun$reduce$1.apply(RDD.scala:1005)
    at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:147)
    at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:108)
    at org.apache.spark.rdd.RDD.withScope(RDD.scala:306)
    at org.apache.spark.rdd.RDD.reduce(RDD.scala:985)
    at com.brkyvz.spark.recommendation.LatentMatrixFactorizationModel$.initialize(LatentMatrixFactorizationModel.scala:116)
    at com.brkyvz.spark.recommendation.StreamingLatentMatrixFactorization$$anonfun$trainOn$1.apply(LatentMatrixFactorization.scala:104)
    at com.brkyvz.spark.recommendation.StreamingLatentMatrixFactorization$$anonfun$trainOn$1.apply(LatentMatrixFactorization.scala:102)
```

Looks like we might need an empty RDD check here:

https://github.com/brkyvz/streaming-matrix-factorization/blob/master/src/main/scala/com/brkyvz/spark/recommendation/LatentMatrixFactorizationModel.scala#L116

what do you think?

perhaps you probably tested this with a file-based stream and missed the edge case where no data yet exists upon the first initialize() vall?  Or perhaps I'm not using the API correctly?

let me know, man.  excited to get this working.  lots of people asking about this

i'm currently trying to inject an initial dummy rating as a workaround.

here's a code snippet:

```
    val ratingsStream = KafkaUtils.createDirectStream[String, String, StringDecoder, StringDecoder](ssc, kafkaParams, topics)

    val streamingMFParams = new LatentMatrixFactorizationParams()
      .setRank(rank)
      .setIter(maxIterations)
      .setLambda(lambdaRegularization)

    val streamingMF = new StreamingLatentMatrixFactorization(streamingMFParams)

    val alsRatingsStream = ratingsStreams.map(message => {
      val tokens = message._2.split(",")
      val userId = tokens(0).trim.toLong
      val itemId = tokens(1).trim.toLong
      val rating = tokens(2).trim.toLong

      Rating(userId, itemId, rating)
    })

    val streamingMFModel = streamingMF.trainOn(alsRatingsStream)
```

**Comment by brkyvz** [comment]
@cfregly Hey! You are correct. A couple of things:
1) I'm perfectly open to contributions, PR's! If you can help me get the streaming side working, I would greatly appreciate it!
2) I don't think we should put the rdd.count check in there, but outside the `initialize` method. Basically, skip `initialize` or `optimizer.train` whenever the RDD is empty.
3) In the current state, I don't think the Streaming side is production ready at all. Instead of using only IndexedRDD, it would be best if it could be backed up by Cassandra or something as well, otherwise, I believe the algorithm will OOM after a day or even a couple of hours at best.

Basically what I'm thinking is, if the user or item vector exists in the IndexedRDD, use it. If not, ask Cassandra. If it doesn't exist there either, initialize a random vector. After some time, if we're not getting any data for a user or item (more likely, user) store it to Cassandra, and free memory from the IndexedRDD.

What do you think?

**Comment by cfregly** [comment]
got this working over in my repo in a more-production-like setting.

here are the relevant links:

Newly-refactored code from this repo:

https://github.com/fluxcapacitor/pipeline/tree/master/myapps/spark/streaming/src/main/scala/com/advancedspark/streaming/ml/incremental

Use of the code above in a Spark Streaming job pull from Kafka:

https://github.com/fluxcapacitor/pipeline/blob/master/myapps/spark/streaming/src/main/scala/com/advancedspark/streaming/rating/ml/TrainMFIncremental.scala

closing this and moving future development to the new repo @ https://github.com/fluxcapacitor/pipeline

thanks so much, burak!
